python_modules/dagster/dagster/_core/definitions/asset_graph_subset.py

```diff
@@ -1,70 +1,129 @@
 import operator
 from collections import defaultdict
 from datetime import datetime
-from typing import AbstractSet, Any, Callable, Dict, Iterable, Mapping, Optional, Set, Union, cast
+from functools import cached_property
+from typing import (
+    AbstractSet,
+    Any,
+    Callable,
+    Dict,
+    Iterable,
+    Mapping,
+    NamedTuple,
+    Optional,
+    Set,
+    Union,
+    cast,
+)
 
 from dagster import _check as check
 from dagster._core.definitions.partition import (
     PartitionsDefinition,
     PartitionsSubset,
 )
+from dagster._core.definitions.time_window_partitions import PartitionKeysTimeWindowPartitionsSubset
 from dagster._core.errors import (
     DagsterDefinitionChangedDeserializationError,
 )
 from dagster._core.instance import DynamicPartitionsStore
+from dagster._serdes.serdes import NamedTupleSerializer, whitelist_for_serdes
 
 from .asset_graph import AssetGraph
 from .events import AssetKey, AssetKeyPartitionKey
 
 
-class AssetGraphSubset:
-    def __init__(
-        self,
-        asset_graph: AssetGraph,
-        partitions_subsets_by_asset_key: Optional[Mapping[AssetKey, PartitionsSubset]] = None,
+class AssetGraphSubsetSerializer(NamedTupleSerializer):
+    def before_pack(self, value: "AssetGraphSubset") -> "AssetGraphSubset":
+        converted_partitions_subsets_by_serialized_asset_key = {}
+        for k, v in value.partitions_subsets_by_serialized_asset_key.items():
+            if isinstance(v, PartitionKeysTimeWindowPartitionsSubset):
+                converted_partitions_subsets_by_serialized_asset_key[
+                    k
+                ] = v.to_time_window_partitions_subset()
+            else:
+                converted_partitions_subsets_by_serialized_asset_key[k] = v
+
+        return value._replace(
+            partitions_subsets_by_serialized_asset_key=converted_partitions_subsets_by_serialized_asset_key
+        )
+
+
+@whitelist_for_serdes(serializer=AssetGraphSubsetSerializer)
+class AssetGraphSubset(
+    NamedTuple(
+        "_AssetGraphSubset",
+        [
+            ("partitions_subsets_by_serialized_asset_key", Mapping[str, PartitionsSubset]),
+            ("non_partitioned_asset_keys", AbstractSet[AssetKey]),
+        ],
+    )
+):
+    def __new__(
+        cls,
+        partitions_subsets_by_serialized_asset_key: Optional[Mapping[str, PartitionsSubset]] = None,
         non_partitioned_asset_keys: Optional[AbstractSet[AssetKey]] = None,
+        partitions_subsets_by_asset_key: Optional[Mapping[AssetKey, PartitionsSubset]] = None,
```

Comment:
adds the `partitions_subsets_by_asset_key` param as a helper param that converts the asset keys into serializable form